src/js/components/conversation.jsx

```diff
@@ -54,34 +66,146 @@ export class ConversationComponent extends Component {
+                this._isScrolling = false;
+                this._forceScrollToBottom = false;
+            });
+            this.scrollTimeouts.push(timeout);
+        }
+    };
+
+    scrollToPreviousFirstMessage = (node = undefined) => {
+        // This will scroll to specified node if we've reached the oldest messages.
+        // Otherwise, scroll to this._lastTopMessageNode
+        if (node) {
             const container = findDOMNode(this);
-            const logo = this.refs.logo;
-            const scrollTop = container.scrollHeight - container.clientHeight - logo.clientHeight - INTRO_BOTTOM_SPACER;
-            container.scrollTop = scrollTop;
-        });
-        this.scrollTimeouts.push(timeout);
+            container.scrollTop = getTop(node, container) - LOAD_MORE_LINK_HEIGHT;
+        } else {
+            if (this._lastTopMessageNodePosition && !this._isScrolling) {
+                const container = findDOMNode(this);
+                const node = this._lastTopMessageNode;
+                this._isScrolling = true;
+
+                // When fetching more messages, we want to make sure that after 
+                // render, the messages stay in the same places
+                container.scrollTop = getTop(node, container) - this._lastTopMessageNodePosition;
+
+                const timeout = setTimeout(() => {
+                    this._isScrolling = false;
+                });
+
+                this.scrollTimeouts.push(timeout);
+            }
+        }
+        this._lastTopMessageNode = undefined;
+        
     };
 
+    componentWillUpdate(nextProps) {
+        const {messages: currentMessages, isFetchingMoreMessages} = this.props;
+        const {messages: newMessages} = nextProps;
+
+        // Check for new appMaker (and whisper) messages
+        const isAppMakerMessage = newMessages.length - currentMessages.length === 1 ? newMessages.slice(-1)[0].role !== 'appUser' : false;
+        if (isAppMakerMessage && !isFetchingMoreMessages) {
+            const container = findDOMNode(this);
+            const appMakerMessageBottom = this._lastMessageNode.getBoundingClientRect().bottom;
+            const containerBottom = container.getBoundingClientRect().bottom;
+
+            // If appMaker message is 'in view', we should scroll to bottom.
+            // Otherwise, don't scroll
+            if (appMakerMessageBottom <= containerBottom) {
+                this._forceScrollToBottom = true;
+            } else {
+                this._forceScrollToBottom = false;
+            }
+        }
+    }
+
     componentDidMount() {
+        // On component render, force scroll to bottom, or else conversation will
+        // find itself at a random spot
         this.scrollToBottom();
     }
 
     componentDidUpdate() {
-        this.scrollToBottom();
+        if (this.props.isFetchingMoreMessages) {
+            if (this.props.hasMoreMessages) {
+                this.scrollToPreviousFirstMessage();
```

Comment:
To make this logc clearer `scrollToPreviousFirstMessage` ought to be checking `this.props.hasMoreMessages` itself